### ISSUE TITLE
Issue with Vestige + Helm

### ISSUE THREAD

**Issue by wtpisaac** [LMMS/lmms#3918]
I've noticed that when trying to use Helm as a VST under Windows, running on LMMS v1.2.0rc4, attempting to open a menu in Helm does not work.

**Expected Behavior:** Clicking on Distortion menu in the Helm synth opens the Distortion menu. Right clicking opens the right click menu.

**Actual behavior:** These menus open, but then close immediately, too quickly to interact with anything in it. _Very rarely_, it will work, which I notice causes Helm's animations to stop.

I believe this is an issue with LMMS (specifically vestige?) as the standalone Helm application does not have this issue on Windows, nor does this issue occur in other DAWs.

**Comment by tresf** [comment]
Tagging @DomClark -- our resident expert with VSTs. :)

**Comment by DomClark** [comment]
Successfully reproduced. I've found one other host where this problem occurs: Hermann Seib's [VSTHost](http://www.hermannseib.com/english/vsthost.htm) when the plugin is running with a different number of bits to that of the host, in which case it is loaded in a separate bridge process, like LMMS does. Presumably having the parent window in a different process messes something up somewhere. I'm quite busy with university work currently, but I'll take a look when I have time. In the meantime, the value can be changed by hovering over the menu and scrolling.

**Comment by DomClark** [comment]
Here is a pair of Helm builds with `Process::isForegroundProcess` patched to always return `true`, which should behave slightly better. Build is of master at time of writing, at commit [81a65d0](https://github.com/mtytel/helm/tree/81a65d0e68980f59450c1e46d008a54784ca6674). Standard "this may break stuff" disclaimers apply.

[helmpatched.zip](https://github.com/LMMS/lmms/files/1424561/helmpatched.zip)

**Comment by zonkmachine** [comment]
> somebody should probably just file an issue with JUCE.

Maybe this is a job for the Helm devs?
JUCE issue tracker here: https://github.com/WeAreROLI/JUCE

**Comment by tresf** [comment]
@DomClark excellent investigation.  Can you please open a bug at JUCE describing the issue?

**Comment by PhysSong** [comment]
Can be reproduced on Windows, but not on Linux. FYI one can work around this issue once #3786 is merged, by using separate VST window mode. Note that this isn't a complete fix...

**Comment by zonkmachine** [comment]
@DomClark Can you please open an issue with JUCE?
https://github.com/WeAreROLI/JUCE
I wouldn't know how to word it.

Also, bumping helm @mtytel

**Comment by DomClark** [comment]
Ah sorry, I completely forgot about this! I'll open an issue there soon.

**Comment by zonkmachine** [comment]
Seem to be a problem with Dexed as well ([forum comment](https://lmms.io/forum/viewtopic.php?f=7&t=27363#p54185)).

**Comment by zonkmachine** [comment]
> Through some internal logic while opening a popup menu, execution reaches MouseSourceState::checkButtonState. From line 1039, the code is
> 
>> ```
>> if (! window.doesAnyJuceCompHaveFocus())
>> {
>>      if (timeNow > window.lastFocusedTime + 10)
>>      {
>>```

The [lines just ahead of this code](https://github.com/WeAreROLI/JUCE/blob/481221a25612d032a33a4fb5c4c1ac20e58e9789/modules/juce_gui_basics/menus/juce_PopupMenu.cpp#L1078:L1080) recently changed. Don't know if it applies to this case though.

**Comment by DomClark** [comment]
I've found a workaround here: if one clears the `WS_CHILD` style from the LVSL window, then `GetForegroundWindow` assumes it's a top-level window and will return its handle when opening the menu, so `isForegroundProcess` behaves as desired and everything works properly. This doesn't have any immediately obvious side effects, but I'm going to test it out for a while as this is somewhat abusing Win32 and may have consequences somewhere.
This also fixes Sylenth's menus as well as #4424.

**Comment by zonkmachine** [comment]
Fine, but please open an issue up-stream. I can't do this out of lack of competence on the matter.

**Comment by DomClark** [comment]
I've had a look at what some other DAWs do and it turns out FL also uses the `WS_CHILD` trick, so it should be safe to include.
JUCE issue opened here: https://github.com/WeAreROLI/JUCE/issues/401.

**Comment by zonkmachine** [comment]
> I've had a look at what some other DAWs do and it turns out FL also uses the `WS_CHILD` trick, so it should be safe to include.

As there seem to be no action in [WeAreROLI/JUCE#401](https://github.com/WeAreROLI/JUCE/issues/401) maybe it's best to fix this on our side as you've suggested?

**Comment by DomClark** [comment]
> Maybe it's best to fix this on our side as you've suggested?

Sure. Although I think this should go into master, not stable-1.2, since Lukas-W wants #2826 in master which seems a more minor change than this.

**Comment by musikBear** [comment]
@DomClark -yet one more VST with odd RC7win32 behavior.
http://www.vst4free.com/free_vst.php?plugin=Marimbaphonic&id=2013 
1) the 'Dynamic' dial resets it self to a different value that it is saved as
2) The LMMS Dial-array is empty. There are no controllers, so automation is impossible.
3) The 'preset-collection' is empty, even thoug a special folder with internal presets are part of the download.
So yet one dumb VST that acts up in LMMS :|

**Comment by zonkmachine** [comment]
> -yet one more VST with odd RC7win32 behavior.

And how is this related to the JUCE bug?
It isn't. Please open a new issue.

**Comment by musikBear** [comment]
Couldent this be a consolidating ticket for VST related issues in specific instruments/ effects?
One separate issue for each of these is overkill, and there may very well be similarities between the issues, then these are less obvious, if they are in different tickets, but that ofcause just my opinion.
If you like i can make a consolidating of everything related to VST-fails, that still are open. I got time..

**Comment by zonkmachine** [comment]
> Couldent this be a consolidating ticket for VST related issues in specific instruments/ effects?

No. This whole thread is one single issue and it's a complex one. Marimbaphonic doesn't look like it's JUCE based. Please open a new issue and I will clean up this discussion.

**Comment by musikBear** [comment]
Will do/ done

**Comment by zonkmachine** [comment]
_This issue has been cleaned from off topic posts_

**Comment by jpcima** [comment]
Hi I'm also impacted jpcima/ADLplug#53 and so is JuceOPLVSTi.
From what I observed, it's only LMMS of the hosts tried which exhibits this sort of problem.
The very latest JUCE from develop was not helpful.

What's the advice for a plugin developer to work around this?
a recommended patch that I can introduce currently in my release builds?

**Comment by DomClark** [comment]
@jpcima This bug should manifest itself in any host which bridges plugins and embeds plugin UIs in the host's main window, unless it specifically works around the issue (e.g. FL Studio). Another such host is [VSTHost](http://www.hermannseib.com/english/vsthost.htm) (when the plugin architecture differs from that of VSTHost), and there may be more that I'm not aware of. We intend to add a workaround, but it won't land in our next stable release.

For a workaround to use in your plugin, you can patch `juce::Process::isForegroundProcess` to always return `true`. It should suffice to only do this for the Win32 version (in `modules/juce_gui_basics/native/juce_win32_Windowing.cpp`), since LMMS (and VSTHost) only currently support Windows VST plugins. I haven't tested this thoroughly for any undesirable side effects, but it does fix the bug. Alternatively, you can instruct users to change the plugin embedding mode in LMMS to "no embedding". This option is available in our latest beta release, and will be in a stable release soon (hopefully).

**Comment by DomClark** [comment]
The following patch works on Windows, but I haven't tested it on Linux yet. PhysSong reports that the bug doesn't occur on Linux, so it may not need to be enabled there anyway. I haven't found any undesirable side effects yet, unlike the approach of not setting `WS_CHILD` on the wrapper.
```diff
diff --git a/plugins/vst_base/RemoteVstPlugin.cpp b/plugins/vst_base/RemoteVstPlugin.cpp
index 942c9e960..121f04b00 100644
--- a/plugins/vst_base/RemoteVstPlugin.cpp
+++ b/plugins/vst_base/RemoteVstPlugin.cpp
@@ -282,6 +282,8 @@ public:
 	// has to be called as soon as input- or output-count changes
 	int updateInOutCount();
 
+	HWND pluginHwnd() const { return m_window; }
+
 	inline void lockShm()
 	{
 		m_shmLock.lock();
@@ -435,6 +437,55 @@ private:
 } ;
 
 
+// Workaround for https://github.com/juce-framework/JUCE/issues/401.
+// JUCE tests if it has focus by comparing its process ID to that of the process
+// to which the currently active window belongs. This is incorrect for bridged,
+// embedded plugins; when the host's window is active, the bridge's process ID
+// won't the match the active window's process ID, i.e. that of the host.
+// We work around this by patching `GetWindowThreadProcessId` to return the
+// bridge's process ID when given a handle to the host's top level window.
+DWORD WINAPI GetWindowThreadProcessIdPatch(HWND hWnd, LPDWORD lpdwProcessId)
+{
+	// Is `hWnd` the top level window that contains the plugin?
+	if (GetAncestor(__plugin->pluginHwnd(), GA_ROOT) == hWnd)
+	{
+		// If so, pretend that it belongs to the current process.
+		if(lpdwProcessId) { *lpdwProcessId = GetCurrentProcessId(); }
+		return GetCurrentThreadId();
+	}
+	// Otherwise use the default behavior.
+	return GetWindowThreadProcessId(hWnd, lpdwProcessId);
+}
+
+// Walk the import table of the plugin and replace `GetWindowThreadProcessId`
+// with our own version.
+void injectJuceMenuWorkaround(HINSTANCE pluginLibrary)
+{
+	const auto base = reinterpret_cast<std::intptr_t>(pluginLibrary);
+	const auto dosHeader = reinterpret_cast<PIMAGE_DOS_HEADER>(base);
+	const auto ntHeaders = reinterpret_cast<PIMAGE_NT_HEADERS>(base + dosHeader->e_lfanew);
+	const auto& importDir = ntHeaders->OptionalHeader.DataDirectory[IMAGE_DIRECTORY_ENTRY_IMPORT];
+	for(auto importDesc = reinterpret_cast<PIMAGE_IMPORT_DESCRIPTOR>(base + importDir.VirtualAddress);
+		importDesc->Name != 0; ++importDesc)
+	{
+		const auto libraryName = reinterpret_cast<LPCSTR>(base + importDesc->Name);
+		if(stricmp(libraryName, "user32.dll") != 0) { continue; }
+		for(auto originalFirstThunk = reinterpret_cast<PIMAGE_THUNK_DATA>(base + importDesc->OriginalFirstThunk),
+				firstThunk = reinterpret_cast<PIMAGE_THUNK_DATA>(base + importDesc->FirstThunk);
+			originalFirstThunk->u1.AddressOfData != 0; ++originalFirstThunk, ++firstThunk)
+		{
+			const auto functionName = reinterpret_cast<PIMAGE_IMPORT_BY_NAME>(
+				base + originalFirstThunk->u1.AddressOfData);
+			if(strcmp(functionName->Name, "GetWindowThreadProcessId") != 0) { continue; }
+			DWORD oldProtect = 0;
+			using FuncType = decltype(firstThunk->u1.Function);
+			VirtualProtect(&firstThunk->u1.Function, sizeof(FuncType), PAGE_READWRITE, &oldProtect);
+			firstThunk->u1.Function = reinterpret_cast<FuncType>(&GetWindowThreadProcessIdPatch);
+			VirtualProtect(&firstThunk->u1.Function, sizeof(FuncType), oldProtect, &oldProtect);
+			return;
+		}
+	}
+}
 
 
 #ifdef SYNC_WITH_SHM_FIFO
@@ -882,6 +933,8 @@ bool RemoteVstPlugin::load( const std::string & _plugin_file )
 		return false;
 	}
 
+	injectJuceMenuWorkaround(m_libInst);
+
 	m_plugin = mainEntry( hostCallback );
 	if( m_plugin == NULL )
 	{
```

**Comment by DomClark** [comment]
I tested Sylenth1 some more. I was able to determine that it uses JUCE, and that its menu bug is caused by the same buggy JUCE code. However, this new patch does not work with it due to the obfuscation used, which makes it impossible to overwrite its import address table (when unpacked in memory, I'm not sure it even meaningfully has one). There are other approaches, such as overwriting the code of `GetWindowThreadProcessId`, or patching the export table of `user32.dll`, but these are more invasive and more complex to implement.

**Comment by DomClark** [comment]
JUCE have fixed the bug on their end: https://github.com/juce-framework/JUCE/issues/401#issuecomment-856914463. Do we still want to try and work around this, or should we assume most plugins will update and embedding can be turned off as a workaround for the few that are still broken?

**Comment by zonkmachine** [comment]
> should we assume most plugins will update and embedding can be turned off as a workaround for the few that are still broken?

No. Helm will probably not update because it has been left by the developer for other projects. I don't know about other projects but if there is a possibility to have a workaround for lmms-1.3 I think it could still be worth it if it's not too much work involved. Even a simple fix that would not work with some plugins as the case with Sylenth1 above. I have unfortunately not had time to test your patch.

**Comment by Yutoob12** [comment]
> I've noticed that when trying to use Helm as a VST under Windows, running on LMMS v1.2.0rc4, attempting to open a menu in Helm does not work.
> 
> **Expected Behavior:** Clicking on Distortion menu in the Helm synth opens the Distortion menu. Right clicking opens the right click menu.
> 
> **Actual behavior:** These menus open, but then close immediately, too quickly to interact with anything in it. _Very rarely_, it will work, which I notice causes Helm's animations to stop.
> 
> I believe this is an issue with LMMS (specifically vestige?) as the standalone Helm application does not have this issue on Windows, nor does this issue occur in other DAWs.

Turn off embedding of VST's in the settings
[Edit] > [Settings] > [General Settings] > "Plugin Embedding" > "No Embedding"
This should fix the problem. Now VST's will be kept in a separate window.

**Comment by musikBear** [comment]
It is possible to get Vestige _MIDI-tables_. Some are supposed to 'replace' the internal MIDI-chip, others are meant to emulate famous MIDI-Tables from hard-synths.
At least two of these does not work properly in LMMS.
1) SYXG VST : https://veg.by/en/projects/syxg50/
2) Edirol roland orchestra 1.03 vst
The _expected behaviour_ would be **General MIDI-table-support** Similar to _SF2 with GeneralUser-MIDI-file_
![image](https://user-images.githubusercontent.com/6662066/177374597-2a5b7a3f-64e8-44e2-a781-631a472e5083.png)
This works reported in **VST-HOST**

Actual Behaviour: 
There is only one patch available (1. Grand-piano)
There are no patch-selector or any other way to access any bank.
There are no Dial-array either (Array opens, but is empty)
![image](https://user-images.githubusercontent.com/6662066/177375981-c1fea40d-526e-4f08-8906-e218ef7cf102.png)

Even though LMMS has SF2, it may be something needed to be worked at.

**Comment by zonkmachine** [comment]
> This is not related to Helm

Correct. It's also not related to JUCE which is what this issue is really about. Please move your post above to a new issue. :+1:

**Comment by zonkmachine** [comment]
> Seem to be a problem with Dexed as well ([forum comment](https://lmms.io/forum/viewtopic.php?f=7&t=27363#p54185)).

This should be fixed in the latest Dexed which used juce 7.0.1 (juce 6.1.0 introduced the fixe).

**Comment by zonkmachine** [comment]
I'm leaning toward closing this as `wontfix` and `upstream`.

Last Helm release was seven years ago and it's maybe better that someone fork and update it instead of us providing workarounds. The other projects we know the JUCE version they build against seem to have updated if they're still active.

**Comment by DomClark** [comment]
Seems reasonable to me. I hope to do more work on embedding bugs in the future, and it's possible this may be fixed as a side effect, but it doesn't need to be a target.
